### PR TITLE
CLDR-13636 Make ExampleGenerator faster, clear cache less

### PR DESCRIPTION
**Pull request by btangmu**
-New RecordingCLDRFile.java, ExampleDependencies.java, ExampleCache.java

-TestExampleGeneratorDependencies: use RecordingCLDRFile, write ExampleDependencies

-ExampleGenerator: use ExampleCache, ExampleCacheItem

-ExampleGenerator: replace INHERITANCE_MARKER with Bailey before get cache, not after

-ExampleGenerator: move some code to new subroutine constructExampleHtml

-ExampleCache: use starredPath for cache key; AVOID_CLEARING_CACHE = true

-Prevent some NullPointerExceptions, do not just catch and ignore them; make SHOW_ERROR true

-Call notifyListeners only if the winning value changed

-Call notifyListeners from voteForValue, not from setValueFromResolver

-Delete unused code: WHAT_DELETE, WHAT_VERIFY, deleteValue, internalDeleteValue

-Clean-up

-Comments

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-13636
- [x] Updated PR title and link in previous line to include Issue number

### CODE REVIEW

**Comment by btangmu** [comment]
Draft for now. Needs more testing and review. Also, there seem to be problems with https://github.com/unicode-org/cldr/pull/396 -- parallel streams hitting code that's not thread-safe, which complicate testing this PR.

**Comment by btangmu** [comment]
It turns out the thread-safety issue for #396 has already been addressed yesterday by https://github.com/unicode-org/cldr/pull/402

**Comment by btangmu** [comment]
Testing went well, so I'm making this ready for review.

File: tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java [hunk 1]
```diff
@@ -1143,7 +1144,12 @@ public synchronized void voteForValue(User user, String distinguishingXpath, Str
                 }
             }
 
+            String oldVal = xmlsource.getValueAtDPath(distinguishingXpath);
```
Comment (macchiati) [inline tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java]: looks like this is unused except when newVal != null, so put under that condition.

File: tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java [hunk 2]
```diff
@@ -1293,84 +1304,6 @@ private void internalSetVoteForValue(User user, String distinguishingXpath, Stri
             stamp.next();
         }
 
-        @Override
```
Comment (macchiati) [inline tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java]: Is this unused? Please explain in a comment on the PR.

File: tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java (hunk 1 above)
Comment (btangmu) [reply tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java]: OK -- sorry I merged before I saw your comment. I'll do that in another PR.

File: tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java (hunk 2 above)
Comment (btangmu) [reply tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java]: Yes, it's unused, unreachable code. Though I don't know the history behind WHAT_DELETE and deleteValue, I suspect their original purpose is now accomplished by Abstain.

File: tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java (hunk 1 above)
Comment (btangmu) [reply tools/cldr-apps/src/org/unicode/cldr/web/STFactory.java]: On further examination, we need to get oldVal before processing the vote, and we don't know whether newVal will be null until after processing the vote, so we need to get oldVal unconditionally. Note that newVal is the winning value after the vote, and isn't necessarily the same as "value", which is what the user voted for. "value" is null if the user abstained, but newVal is only null if there's no winning value, not even an inherited or baseline value. This function is only called once, when a user submits a vote, so it's not really a hot spot or inner-loop context. OK?

**Comment by macchiati** [comment]
ok
Mark

**Pull request merged**